src/test/test-dirent-util.c

```diff
@@ -0,0 +1,208 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <dirent.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <sys/stat.h>
+
+#include "alloc-util.h"
+#include "dirent-util.h"
+#include "fs-util.h"
+#include "mkdir.h"
+#include "path-util.h"
+#include "rm-rf.h"
+#include "stat-util.h"
+#include "string-util.h"
+#include "tests.h"
+
+
+
+TEST (test_dirent_ensure_type) {
+        int result;
+        static struct dirent de = {
+                .d_type = DT_UNKNOWN,
+                .d_name = "test",
+        };
+
+        assert_se(de.d_type == DT_UNKNOWN);
+        int dir_fd;
+
+        dir_fd = 0;
+        dirent_ensure_type(dir_fd, &de);
+
+        /* Test when d_name is "." or ".." */
+        strcpy(de.d_name, ".");
+        result = dirent_ensure_type(dir_fd, &de);
+        assert_se(result == 0);
+        assert_se(de.d_type == DT_DIR);
+
+        strcpy(de.d_name, "..");
+        result = dirent_ensure_type(dir_fd, &de);
+        assert_se(result == 0);
+        assert_se(de.d_type == DT_DIR);
+}
+
+TEST (test_dirent_is_file) {
+        static const struct dirent de_unknown = {
+                .d_type = DT_UNKNOWN,
+                .d_name = "test_unknown",
+        };
+
+        static const char *arg_test_dir = NULL;
+
+        const char *tempdir, *name, *dotfile, *name_alias, *bakfile, *tilda;
+        DIR *dir;
+
+        tempdir = strjoina(arg_test_dir ?: "/tmp", "/test-dirent_is_file");
+        name = strjoina(tempdir, "/test.txt");
+        dotfile = strjoina(tempdir, "/.hidden_file");
+        bakfile = strjoina(tempdir, "/test.bak");
+        tilda = strjoina(tempdir, "/test~");
+        name_alias = strjoina(arg_test_dir ?: "/tmp", "/test-dirent_is_file/test_link");
+
+        assert_se(mkdir_safe(tempdir, 0755, getuid(), getgid(), MKDIR_WARN_MODE) >= 0);
+        assert_se(touch(name) >= 0);
+        assert_se(touch(dotfile) >= 0);
+        assert_se(touch(bakfile) >= 0);
+        assert_se(touch(tilda) >= 0);
+
+
+        if (symlink(name, name_alias) < 0) {
+                assert_se(IN_SET(errno, EINVAL, ENOSYS, ENOTTY, EPERM));
+                log_tests_skipped_errno(errno, "symlink() not possible");}
+
+        dir = opendir(tempdir);
+        if (dir == NULL) {
+                perror("opendir");
```

Comment:
don't use perror, use logging functions

Reply:
can you kindly point out the logging functions ?, thank you.

Reply:
I found them.